opensearch/src/main/java/org/opensearch/sql/opensearch/data/value/OpenSearchExprValueFactory.java

```diff
@@ -526,4 +527,26 @@ private ExprValue parseInnerArrayValue(
   private String makeField(String path, String field) {
     return path.equalsIgnoreCase(TOP_PATH) ? field : String.join(".", path, field);
   }
+
+  /**
+   * Recursively extracts the final primitive value from nested Map structures. For example:
+   * {attributes={telemetry={sdk={language=java}}}} -> "java"
+   *
+   * @param value The value to extract from
+   * @return The extracted primitive value, or the original value if extraction is not possible
+   */
+  @SuppressWarnings("unchecked")
+  private Object extractFinalPrimitiveValue(Object value) {
+    if (value == null || !(value instanceof Map)) {
+      return value;
+    }
+
+    Map<String, Object> map = (Map<String, Object>) value;
```

Comment:
Why do this cast at runtime instead of just specifying the type in the signature and pushing the check to the left?